### PR TITLE
Auto save Feature

### PR DESCRIPTION
**Pull request by Sanjay-B**
When enabled, your place is saved every five seconds. I felt that five seconds is reasonable enough.

### CODE REVIEW

**Comment by sasial-roblox** [comment]
Wouldn't you want to go like every minute so it doesn't lag with huge worlds?
Even every 5?
OR, I can make an option to choose how often.

**Comment by Sanjay-B** [comment]
I thought about it every extensively. Long term, the time limit is going to be editable in under the general tab in settings. For now, five seconds is long enough to not completely spam but, to ensure that progress is saved.

A lot of work can get done in a minute and could be potentially missed. This could anger a lot of end users down the road which is what we don’t want. For now, it defaults to five seconds until I explicitly add it as a property in settings.

File: workshop/controllers/ui/components/autoSave.lua [hunk 1]
```diff
@@ -0,0 +1,17 @@
+-- Auto Save 
+local shared = require("tevgit:workshop/controllers/shared.lua")
+local autoSave = {}
+
+autoSave.Enabled = false
+autoSave.timeLimit = 5 -- Default (move to settings value soon)
+
+autoSave.Sync = spawnThread(function()
```
Comment (teverse) [inline workshop/controllers/ui/components/autoSave.lua]: once this thread is killed, autosave is killed and cannot be toggled back on

File: workshop/controllers/ui/components/settings.lua [hunk 2]
```diff
@@ -120,6 +120,19 @@ if shared.developerMode then
...
 
+	-- Auto Save / Sync
+	local syncThread
+	local autoSaveToggle = ui.button(developmentPage, autoSave.Enabled and "Disable Auto-Save" or "Enabled Auto-Save", guiCoord(0, 190, 0, 30), guiCoord(0, 15, 0, 290), "secondary")
+	autoSaveToggle:mouseLeftPressed(function()
+		autoSave.Enabled = not autoSave.Enabled
+		if autoSave.Enabled then
+			syncThread = autoSave.Sync -- create syncThread
+		else
+			syncThread:kill() -- kill syncThread
```
Comment (teverse) [inline workshop/controllers/ui/components/settings.lua]: consider moving autosave logic that deals with autosave threads to the autosave module

File: workshop/controllers/ui/components/autoSave.lua (hunk 1 above)
Comment (Sanjay-B) [reply workshop/controllers/ui/components/autoSave.lua]: Meant to wrap this in a ``function() end)``. That's on me.

Comment (Sanjay-B) [reply workshop/controllers/ui/components/autoSave.lua]: Will be fixed with the next commit to this pull.

File: workshop/controllers/ui/components/settings.lua (hunk 2 above)
Comment (Sanjay-B) [reply workshop/controllers/ui/components/settings.lua]: Again, that's on me. Will be fixed in the next commit.

**Comment by Sanjay-B** [comment]
It should be fine now.

File: workshop/controllers/ui/components/autoSave.lua [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+        while true do
+            if autoSave.Enabled then
+                shared.workshop:saveGame()
+            end
+            wait(autoSave.timeLimit)
+        end
+    end)
+end
+
+return autoSave
```
Comment (sasial-roblox) [inline workshop/controllers/ui/components/autoSave.lua]: As with format, just leave an extra line at the end.
image

Comment (sasial-roblox) [reply workshop/controllers/ui/components/autoSave.lua]: Might just be my OCD, but like yeah

**Comment by Sanjay-B** [comment]
Will do. For now, it's only locked to developers.

**Pull request merged**